# Retail Allocation & Stock Flow Analytics Project

## Step 01 Data Validation

### Objective

Validate data quality and confirm that the dataset is suitable for inventory allocation, replenishment planning and stock flow analysis.

### 1.1 Load Dataset

In [7]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/supply_chain_dataset1.csv")

df.head()

,Date,SKU_ID,Warehouse_ID,Supplier_ID,Region,Units_Sold,Inventory_Level,Supplier_Lead_Time_Days,Reorder_Point,Order_Quantity,Unit_Cost,Unit_Price,Promotion_Flag,Stockout_Flag,Demand_Forecast
0,2024-01-01,SKU_1,WH_1,SUP_8,West,10,592,14,379,0,13.95,20.48,0,0,8.52
1,2024-01-02,SKU_1,WH_1,SUP_8,West,17,575,14,379,0,13.95,20.48,0,0,18.63
2,2024-01-03,SKU_1,WH_1,SUP_8,North,35,540,14,379,0,13.95,20.48,1,0,39.62
3,2024-01-04,SKU_1,WH_1,SUP_8,South,24,516,14,379,0,13.95,20.48,0,0,19.43
4,2024-01-05,SKU_1,WH_1,SUP_8,West,21,495,14,379,0,13.95,20.48,0,0,18.70


### 1.2 Dataset Overview

In [4]:
df.shape

(91250, 15)

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 91250 entries, 0 to 91249
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Date                     91250 non-null  str    
 1   SKU_ID                   91250 non-null  str    
 2   Warehouse_ID             91250 non-null  str    
 3   Supplier_ID              91250 non-null  str    
 4   Region                   91250 non-null  str    
 5   Units_Sold               91250 non-null  int64  
 6   Inventory_Level          91250 non-null  int64  
 7   Supplier_Lead_Time_Days  91250 non-null  int64  
 8   Reorder_Point            91250 non-null  int64  
 9   Order_Quantity           91250 non-null  int64  
 10  Unit_Cost                91250 non-null  float64
 11  Unit_Price               91250 non-null  float64
 12  Promotion_Flag           91250 non-null  int64  
 13  Stockout_Flag            91250 non-null  int64  
 14  Demand_Forecast          91250 no

### 1.3 Missing Value Check

In [6]:
df.isnull().sum()

Date                       0
SKU_ID                     0
Warehouse_ID               0
Supplier_ID                0
Region                     0
Units_Sold                 0
Inventory_Level            0
Supplier_Lead_Time_Days    0
Reorder_Point              0
Order_Quantity             0
Unit_Cost                  0
Unit_Price                 0
Promotion_Flag             0
Stockout_Flag              0
Demand_Forecast            0
dtype: int64

### 1.3 Findings

No missing values were identified across all 15 variables. The dataset is complete and suitable for further inventory allocation and replenishment analysis.

### 1.4 Duplicate Record Check

In [8]:
df.duplicated().sum()

np.int64(0)

### 1.4 Findings

No duplicate records were identified in the dataset. This indicates that inventory, sales and replenishment metrics are not inflated by repeated observations.

### 1.5 Date Validation

In [9]:
df["Date"] = pd.to_datetime(df["Date"])

df["Date"].min(), df["Date"].max()

(Timestamp('2024-01-01 00:00:00'), Timestamp('2024-12-30 00:00:00'))

### 1.6 Date Continuity Check

In [ ]:
df["Date"].nunique()

365

### 1.5 Findings

The dataset covers a full year from 2024-01-01 to 2024-12-30 and contains 365 unique dates. The date series is complete and suitable for time-based inventory, replenishment and allocation analysis.

### 1.6 Entity Validation

In [ ]:
df["SKU_ID"].nunique()

50

In [12]:
df["Warehouse_ID"].nunique()

5

In [13]:
df["Supplier_ID"].nunique()

10

In [14]:
df["Region"].nunique()

4

In [15]:
pd.DataFrame({
    "Entity": ["SKU", "Warehouse", "Supplier", "Region"],
    "Count": [
        df["SKU_ID"].nunique(),
        df["Warehouse_ID"].nunique(),
        df["Supplier_ID"].nunique(),
        df["Region"].nunique()
    ]
})

,Entity,Count
0,SKU,50
1,Warehouse,5
2,Supplier,10
3,Region,4


### 1.6 Findings

The dataset contains:

- 50 unique SKUs
- 5 warehouses
- 10 suppliers
- 4 regions

This structure supports allocation and stock flow analysis across multiple products, locations and supply sources. The presence of warehouse and regional dimensions is particularly valuable for evaluating right-product-right-place allocation decisions.

### 1.7 Numeric Field Validation

In [16]:
numeric_cols = [
    "Units_Sold",
    "Inventory_Level",
    "Supplier_Lead_Time_Days",
    "Reorder_Point",
    "Order_Quantity",
    "Unit_Cost",
    "Unit_Price",
    "Demand_Forecast"
]

df[numeric_cols].describe()

,Units_Sold,Inventory_Level,Supplier_Lead_Time_Days,Reorder_Point,Order_Quantity,Unit_Cost,Unit_Price,Demand_Forecast
count,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000,91250.000000
mean,20.054564,471.522312,7.984000,300.068000,19.272493,12.203320,18.261800,20.082033
std,9.068602,133.488002,3.907929,54.879945,82.340831,4.574982,7.121136,9.503955
min,0.000000,168.000000,2.000000,201.000000,0.000000,5.020000,6.950000,0.000000
25%,13.000000,370.000000,4.000000,252.000000,0.000000,8.180000,12.000000,12.950000
50%,20.000000,461.000000,8.000000,300.000000,0.000000,11.990000,18.180000,19.950000
75%,27.000000,564.000000,11.000000,346.000000,0.000000,16.320000,23.390000,26.930000
max,59.000000,990.000000,14.000000,398.000000,499.000000,19.760000,35.100000,61.420000


### 1.7 Findings

Key observations from the numeric fields include:

- Inventory levels range from 168 to 990 units, indicating relatively high stock availability across warehouses.
- Supplier lead times vary from 2 to 14 days, suggesting different replenishment requirements across suppliers.
- Reorder points range from 201 to 398 units and appear to drive replenishment decisions.
- Most records have an order quantity of zero, indicating that replenishment events occur only when inventory reaches specific thresholds.
- Demand forecasts closely align with sales volumes, supporting future forecast accuracy and allocation analysis.

These findings suggest the dataset is well suited for inventory coverage, replenishment planning and allocation decision analysis.

# Step 02 Inventory Coverage Analysis

### Objective

Evaluate how long current inventory can support forecast demand and identify potential inventory imbalance across products and warehouses.

### 2.1 Calculate Inventory Coverage

In [17]:
df["Inventory_Coverage"] = (
    df["Inventory_Level"]
    /
    df["Demand_Forecast"]
)

df["Inventory_Coverage"].describe()

/Users/clorisyu/Documents/招聘6.1/macpac/Retail Allocation & Stock Flow Analytics Project/venv/lib/python3.12/site-packages/pandas/core/nanops.py:1028: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)


count    9.125000e+04
mean              inf
std               NaN
min      3.719599e+00
25%      1.617983e+01
50%      2.364629e+01
75%      3.766648e+01
max               inf
Name: Inventory_Coverage, dtype: float64

In [19]:
(df["Demand_Forecast"] == 0).sum()
df[df["Demand_Forecast"] == 0]

,Date,SKU_ID,Warehouse_ID,Supplier_ID,Region,Units_Sold,Inventory_Level,Supplier_Lead_Time_Days,Reorder_Point,Order_Quantity,Unit_Cost,Unit_Price,Promotion_Flag,Stockout_Flag,Demand_Forecast,Inventory_Coverage
240,2024-08-28,SKU_1,WH_1,SUP_8,West,0,449,14,379,0,13.95,20.48,0,0,0.0,inf
258,2024-09-15,SKU_1,WH_1,SUP_8,North,1,481,14,379,0,13.95,20.48,0,0,0.0,inf
294,2024-10-21,SKU_1,WH_1,SUP_8,South,1,578,14,379,0,13.95,20.48,1,0,0.0,inf
642,2024-10-04,SKU_1,WH_2,SUP_5,West,0,432,8,322,0,11.46,19.92,1,0,0.0,inf
648,2024-10-10,SKU_1,WH_2,SUP_5,West,0,377,8,322,0,11.46,19.92,0,0,0.0,inf
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
90794,2024-10-01,SKU_50,WH_4,SUP_8,East,0,300,14,218,0,18.62,22.63,0,0,0.0,inf
90829,2024-11-05,SKU_50,WH_4,SUP_8,East,4,292,14,218,0,18.62,22.63,0,0,0.0,inf
91088,2024-07-22,SKU_50,WH_5,SUP_10,North,3,462,7,283,0,6.65,9.60,0,0,0.0,inf
91142,2024-09-14,SKU_50,WH_5,SUP_10,East,0,397,7,283,0,6.65,9.60,0,0,0.0,inf


In [20]:
(df["Demand_Forecast"] == 0).sum()

np.int64(892)

### 2.2 Handle Zero Forecast Records

In [21]:
df["Inventory_Coverage_Clean"] = np.where(
    df["Demand_Forecast"] > 0,
    df["Inventory_Level"] / df["Demand_Forecast"],
    np.nan
)
df["Inventory_Coverage_Clean"].describe()

count    90358.000000
mean        45.134360
std        477.181086
min          3.719599
25%         16.118469
50%         23.486595
75%         37.021476
max      74900.000000
Name: Inventory_Coverage_Clean, dtype: float64

### 2.3 Investigate Extreme Coverage Values

In [22]:
df.sort_values(
    by="Inventory_Coverage_Clean",
    ascending=False
)[[
    "SKU_ID",
    "Warehouse_ID",
    "Inventory_Level",
    "Demand_Forecast",
    "Inventory_Coverage_Clean"
]].head(10)

,SKU_ID,Warehouse_ID,Inventory_Level,Demand_Forecast,Inventory_Coverage_Clean
7531,SKU_5,WH_1,749,0.01,74900.0
11559,SKU_7,WH_2,641,0.01,64100.0
18821,SKU_11,WH_2,442,0.01,44200.0
5421,SKU_3,WH_5,406,0.01,40600.0
81291,SKU_45,WH_3,612,0.02,30600.0
8273,SKU_5,WH_3,602,0.02,30100.0
18168,SKU_10,WH_5,286,0.01,28600.0
2467,SKU_2,WH_2,523,0.02,26150.0
13484,SKU_8,WH_2,520,0.02,26000.0
75782,SKU_42,WH_3,603,0.03,20100.0


### 2.4 Inventory Coverage Distribution

In [23]:
df["Inventory_Coverage_Clean"].quantile(
    [0.25, 0.5, 0.75, 0.9, 0.95, 0.99]
)

0.25     16.118469
0.50     23.486595
0.75     37.021476
0.90     61.454279
0.95     90.545488
0.99    268.819838
Name: Inventory_Coverage_Clean, dtype: float64

### 2.5 Inventory Health Classification

In [24]:
conditions = [
    df["Inventory_Coverage_Clean"] < 14,

    (df["Inventory_Coverage_Clean"] >= 14) &
    (df["Inventory_Coverage_Clean"] < 30),

    (df["Inventory_Coverage_Clean"] >= 30) &
    (df["Inventory_Coverage_Clean"] <= 90),

    df["Inventory_Coverage_Clean"] > 90
]

choices = [
    "Critical",
    "Healthy",
    "High Inventory",
    "Overstock"
]

df["Inventory_Status"] = np.select(
    conditions,
    choices,
    default="Unknown"
)

df["Inventory_Status"].value_counts()

Inventory_Status
Healthy           43262
High Inventory    27224
Critical          15305
Overstock          4567
Unknown             892
Name: count, dtype: int64

### 2.6 Inventory Health Distribution

In [25]:
round(
    df["Inventory_Status"]
    .value_counts(normalize=True)
    * 100,
    2
)

Inventory_Status
Healthy           47.41
High Inventory    29.83
Critical          16.77
Overstock          5.00
Unknown            0.98
Name: proportion, dtype: float64

### Step 02 Findings

- 47.4% of inventory records were classified as Healthy, indicating that a large proportion of stock is aligned with forecast demand.

- 16.8% of inventory records were classified as Critical, suggesting potential replenishment risk and requiring closer monitoring.

- 34.8% of inventory records were classified as either High Inventory or Overstock, highlighting potential opportunities for stock reallocation and inventory optimisation.

- Inventory Coverage proved to be a more meaningful measure of stock health than Inventory Level alone because it considered both stock availability and forecast demand.

- Extreme coverage values were primarily driven by near-zero demand forecasts rather than unusually high inventory levels, indicating potential slow-moving inventory.


# Step 03 Warehouse Performance Analysis

### Objective

Evaluate inventory health across warehouses to identify where stock flow, replenishment attention and allocation actions may be required.

### 3.1 Warehouse Inventory Status Summary

In [26]:
warehouse_status_count = pd.crosstab(
    df["Warehouse_ID"],
    df["Inventory_Status"]
)

warehouse_status_count

Inventory_Status,Critical,Healthy,High Inventory,Overstock,Unknown
Warehouse_ID,,,,,
WH_1,3160,8759,5313,854,164
WH_2,2654,8656,5783,973,184
WH_3,3434,8523,5205,910,178
WH_4,2971,8618,5503,975,183
WH_5,3086,8706,5420,855,183


### 3.2 Warehouse Inventory Status Percentage

In [27]:
warehouse_status_pct = pd.crosstab(
    df["Warehouse_ID"],
    df["Inventory_Status"],
    normalize="index"
) * 100

warehouse_status_pct.round(2)

Inventory_Status,Critical,Healthy,High Inventory,Overstock,Unknown
Warehouse_ID,,,,,
WH_1,17.32,47.99,29.11,4.68,0.90
WH_2,14.54,47.43,31.69,5.33,1.01
WH_3,18.82,46.70,28.52,4.99,0.98
WH_4,16.28,47.22,30.15,5.34,1.00
WH_5,16.91,47.70,29.70,4.68,1.00


### 3.3 Warehouse Coverage Comparison

In [28]:
warehouse_coverage = (
    df.groupby("Warehouse_ID")["Inventory_Coverage_Clean"]
    .median()
    .sort_values(ascending=False)
)

warehouse_coverage

Warehouse_ID
WH_2    24.458345
WH_4    23.821549
WH_5    23.411885
WH_1    23.013965
WH_3    22.765030
Name: Inventory_Coverage_Clean, dtype: float64

### Step 03 Findings

- Warehouse 2 recorded the highest median inventory coverage (24.46 days) and the lowest proportion of critical inventory (14.54%), indicating a more conservative inventory position.

- Warehouse 3 recorded the lowest median inventory coverage (22.77 days) and the highest proportion of critical inventory (18.82%), suggesting greater replenishment risk.

- Inventory health distribution was broadly consistent across warehouses, indicating that inventory performance differences were relatively moderate rather than extreme.

- The comparison highlights the trade-off between inventory availability and inventory efficiency. Warehouses carrying higher inventory coverage experienced lower stock risk, while leaner inventory positions were associated with higher critical inventory levels.


# Step 04 SKU Risk Analysis

### Objective

Identify which SKUs contribute most to critical inventory risk and overstock exposure, supporting product-level allocation and replenishment decisions.

### 4.1 SKU Inventory Status Summary

In [29]:
sku_status_count = pd.crosstab(
    df["SKU_ID"],
    df["Inventory_Status"]
)

sku_status_count.head()

Inventory_Status,Critical,Healthy,High Inventory,Overstock,Unknown
SKU_ID,,,,,
SKU_1,212,870,634,89,20
SKU_10,323,886,521,75,20
SKU_11,250,838,626,89,22
SKU_12,322,846,552,92,13
SKU_13,296,849,570,99,11


### 4.2 Top Critical SKUs

In [30]:
critical_sku = (
    sku_status_count["Critical"]
    .sort_values(ascending=False)
)

critical_sku.head(10)

SKU_ID
SKU_50    476
SKU_44    418
SKU_3     390
SKU_30    388
SKU_45    378
SKU_35    364
SKU_29    360
SKU_34    356
SKU_14    352
SKU_31    350
Name: Critical, dtype: int64

### 4.3 Critical SKU Investigation

In [31]:
top_critical_skus = [
    "SKU_50",
    "SKU_44",
    "SKU_3",
    "SKU_30",
    "SKU_45"
]

df[
    df["SKU_ID"].isin(top_critical_skus)
].groupby("SKU_ID")[[
    "Units_Sold",
    "Demand_Forecast",
    "Inventory_Level",
    "Supplier_Lead_Time_Days"
]].median().round(2)

,Units_Sold,Demand_Forecast,Inventory_Level,Supplier_Lead_Time_Days
SKU_ID,,,,
SKU_3,20.0,20.27,430.0,10.0
SKU_30,20.0,20.11,429.0,4.0
SKU_44,20.0,20.08,418.0,7.0
SKU_45,20.0,19.81,432.0,7.0
SKU_50,20.0,20.14,392.0,10.0


### 4.3 Critical SKU Root Cause Analysis

### Findings

The highest-risk SKUs did not exhibit unusually high demand forecasts or exceptionally long supplier lead times.

Median demand forecasts were consistent with the overall dataset (~20 units), and lead times ranged from 4 to 10 days, which was within the normal operating range.

The primary difference was inventory level. High-risk SKUs maintained lower median inventory positions (392–432 units) compared with the overall inventory median (461 units), resulting in lower inventory coverage and a higher likelihood of entering critical inventory status.

### 4.4 Critical Inventory Rate by SKU

In [32]:
critical_rate = (
    sku_status_count["Critical"]
    / sku_status_count.sum(axis=1)
    * 100
).sort_values(ascending=False)

critical_rate.head(10)

SKU_ID
SKU_50    26.082192
SKU_44    22.904110
SKU_3     21.369863
SKU_30    21.260274
SKU_45    20.712329
SKU_35    19.945205
SKU_29    19.726027
SKU_34    19.506849
SKU_14    19.287671
SKU_31    19.178082
dtype: float64

### 4.5 Top Overstock SKU

In [33]:
overstock_rate = (
    sku_status_count["Overstock"]
    / sku_status_count.sum(axis=1)
    * 100
).sort_values(ascending=False)

overstock_rate.head(10)

SKU_ID
SKU_41    6.082192
SKU_17    6.082192
SKU_21    5.972603
SKU_42    5.863014
SKU_16    5.863014
SKU_29    5.808219
SKU_37    5.698630
SKU_27    5.643836
SKU_48    5.589041
SKU_19    5.589041
dtype: float64

### 4.6 SKU Profile Comparison

In [35]:
critical_rate.sort_values().head()


SKU_ID
SKU_16     9.479452
SKU_41    11.123288
SKU_47    11.561644
SKU_1     11.616438
SKU_17    12.438356
dtype: float64

In [37]:
critical_rate.median()

np.float64(16.876712328767123)

### 4.7 High-Risk vs Low-Risk SKU Comparison

In [38]:
sku_compare = df[
    df["SKU_ID"].isin(["SKU_50", "SKU_16"])
]

sku_compare.groupby("SKU_ID")[[
    "Units_Sold",
    "Demand_Forecast",
    "Inventory_Level",
    "Supplier_Lead_Time_Days",
    "Inventory_Coverage_Clean"
]].median().round(2)

,Units_Sold,Demand_Forecast,Inventory_Level,Supplier_Lead_Time_Days,Inventory_Coverage_Clean
SKU_ID,,,,,
SKU_16,20.0,19.84,512.0,6.0,25.96
SKU_50,20.0,20.14,392.0,10.0,20.45


### 4.7 Findings

A comparison between SKU_50 (highest critical risk) and SKU_16 (lowest critical risk) showed that demand and sales levels were broadly similar.

The key difference was inventory position. SKU_50 maintained a median inventory level of 392 units compared with 512 units for SKU_16, resulting in lower inventory coverage (20.45 vs 25.96 days).

SKU_50 also had a longer supplier lead time (10 vs 6 days), which may further increase replenishment risk if inventory levels are not adjusted appropriately.

These findings suggest that inventory allocation levels, rather than demand patterns, are the primary driver of critical inventory risk.

### 4.8 SKU Risk Segmentation

In [39]:
sku_risk_segment = pd.DataFrame({
    "Critical_Rate": critical_rate
})

conditions = [
    sku_risk_segment["Critical_Rate"] > 20,
    sku_risk_segment["Critical_Rate"] < 14
]

choices = [
    "High Risk",
    "Low Risk"
]

sku_risk_segment["SKU_Risk_Segment"] = np.select(
    conditions,
    choices,
    default="Normal Risk"
)

sku_risk_segment.head()

,Critical_Rate,SKU_Risk_Segment
SKU_ID,,
SKU_50,26.082192,High Risk
SKU_44,22.904110,High Risk
SKU_3,21.369863,High Risk
SKU_30,21.260274,High Risk
SKU_45,20.712329,High Risk


In [40]:
sku_risk_segment["SKU_Risk_Segment"].value_counts()

SKU_Risk_Segment
Normal Risk    36
Low Risk        9
High Risk       5
Name: count, dtype: int64

### Step 04 Findings

- Inventory risk was concentrated in a small subset of products.
- Five SKUs were classified as High Risk, representing 10% of the product portfolio.
- SKU_50 recorded the highest critical inventory rate (26.1%), approximately 55% higher than the portfolio average.
- Root cause analysis indicated that lower inventory positions, rather than unusually high demand, were the primary driver of critical inventory risk.
- Overstock risk was more evenly distributed across products, suggesting that stock shortage risk was a more significant challenge than excess inventory.

### 4.9 Critical Risk Trend Analysis

#### Objective

Investigate when critical inventory events occur and identify potential seasonality or recurring risk patterns.


#### 4.9.1 Create Month Field

In [42]:
df["Month"] = df["Date"].dt.month

df[["Date", "Month"]].head()

,Date,Month
0,2024-01-01,1
1,2024-01-02,1
2,2024-01-03,1
3,2024-01-04,1
4,2024-01-05,1


#### 4.9.2 Monthly Critical Inventory Trend

In [43]:
monthly_critical = (
    df[df["Inventory_Status"] == "Critical"]
    .groupby("Month")
    .size()
)

monthly_critical

Month
1     1050
2     2340
3     3170
4     3101
5     2527
6     1414
7      569
8      195
9       77
10      85
11     206
12     571
dtype: int64

#### 4.9.3 Monthly Inventory Coverage Trend

In [44]:
monthly_coverage = (
    df.groupby("Month")["Inventory_Coverage_Clean"]
    .median()
)

monthly_coverage

Month
1     22.834912
2     16.805311
3     15.298432
4     15.128270
5     16.799420
6     20.081678
7     26.566456
8     36.487570
9     44.704364
10    45.323693
11    35.674276
12    26.728378
Name: Inventory_Coverage_Clean, dtype: float64

#### 4.9.4 Replenishment Trend Analysis

Objective

Investigate whether changes in inventory coverage were driven by replenishment activity.

In [45]:
monthly_orders = (
    df.groupby("Month")["Order_Quantity"]
    .sum()
)

monthly_orders

Month
1     107094
2     198196
3     232052
4     218812
5     210908
6     175040
7     139013
8      96447
9      76988
10     76296
11    102834
12    124935
Name: Order_Quantity, dtype: int64

#### 4.9.5 Monthly Demand Forecast Trend

Objective

Investigate whether changes in inventory coverage were influenced by changes in forecast demand.

In [46]:
monthly_forecast = (
    df.groupby("Month")["Demand_Forecast"]
    .median()
)

monthly_forecast

Month
1     22.585
2     27.090
3     29.690
4     29.450
5     26.920
6     22.535
7     17.290
8     12.700
9     10.200
10    10.160
11    13.030
12    17.420
Name: Demand_Forecast, dtype: float64

### 4.9 Trend Analysis Findings

Monthly analysis revealed a strong relationship between demand forecast, inventory coverage, and critical inventory risk.

Critical inventory events peaked in March and April, when median forecast demand reached approximately 30 units and inventory coverage fell to around 15 days.

Although replenishment activity was also highest during this period, elevated demand continued to place pressure on inventory availability.

As forecast demand declined during the second half of the year, inventory coverage increased substantially and critical inventory events fell sharply.

These findings suggest that forecast demand was the primary driver of inventory risk, highlighting the importance of proactive allocation planning and demand forecasting.

# Step 05 Allocation Recommendation Framework

### Objective

Translate inventory health, SKU risk and demand trend findings into actionable allocation and replenishment recommendations.



In [ ]:

### 5.1 Define Allocation Action 

In [50]:

Summaryaction_map = {
    "Critical": "Increase Allocation / Replenish",
    "Healthy": "Maintain Current Allocation",
    "High Inventory": "Monitor Inventory",
    "Overstock": "Review / Reallocate Stock",
    "Unknown": "Review Forecast"
}

df["Allocation_Action"] = df["Inventory_Status"].map(action_map)

df[[
    "SKU_ID",
    "Warehouse_ID",
    "Region",
    "Inventory_Status",
    "Allocation_Action"
]].head()

,SKU_ID,Warehouse_ID,Region,Inventory_Status,Allocation_Action
0,SKU_1,WH_1,West,High Inventory,Monitor Inventory
1,SKU_1,WH_1,West,High Inventory,Monitor Inventory
2,SKU_1,WH_1,North,Critical,Increase Allocation / Replenish
3,SKU_1,WH_1,South,Healthy,Maintain Current Allocation
4,SKU_1,WH_1,West,Healthy,Maintain Current Allocation


### 5.2 Allocation Action 

In [48]:
action_summary = df["Allocation_Action"].value_counts()

action_summary

Allocation_Action
Maintain Current Allocation        43262
Monitor Inventory                  27224
Increase Allocation / Replenish    15305
Review / Reallocate Stock           4567
Review Forecast                      892
Name: count, dtype: int64

In [49]:
action_summary_pct = (
    df["Allocation_Action"]
    .value_counts(normalize=True)
    * 100
).round(2)

action_summary_pct

Allocation_Action
Maintain Current Allocation        47.41
Monitor Inventory                  29.83
Increase Allocation / Replenish    16.77
Review / Reallocate Stock           5.00
Review Forecast                     0.98
Name: proportion, dtype: float64

# Step 06 Power BI Data Export

Export cleaned and enriched analysis tables for Power BI dashboard development on Windows.

In [52]:
import os

output_folder = "../powerbi_data"
os.makedirs(output_folder, exist_ok=True)

In [53]:
df_powerbi = df.merge(
    sku_risk_segment[["Critical_Rate", "SKU_Risk_Segment"]],
    left_on="SKU_ID",
    right_index=True,
    how="left"
)

df_powerbi.to_csv(
    "../powerbi_data/fact_inventory_enriched.csv",
    index=False
)

In [54]:
sku_risk_segment.to_csv(
    "../powerbi_data/sku_risk_summary.csv"
)

In [55]:
warehouse_summary = (
    df.groupby("Warehouse_ID")
    .agg(
        Median_Coverage=("Inventory_Coverage_Clean", "median"),
        Critical_Rate=("Inventory_Status", lambda x: (x == "Critical").mean() * 100),
        Overstock_Rate=("Inventory_Status", lambda x: (x == "Overstock").mean() * 100),
        Total_Units_Sold=("Units_Sold", "sum"),
        Median_Forecast=("Demand_Forecast", "median")
    )
    .round(2)
    .reset_index()
)

warehouse_summary.to_csv(
    "../powerbi_data/warehouse_summary.csv",
    index=False
)

In [56]:
monthly_trend_summary = (
    df.groupby("Month")
    .agg(
        Critical_Count=("Inventory_Status", lambda x: (x == "Critical").sum()),
        Median_Coverage=("Inventory_Coverage_Clean", "median"),
        Median_Forecast=("Demand_Forecast", "median"),
        Total_Order_Quantity=("Order_Quantity", "sum"),
        Total_Units_Sold=("Units_Sold", "sum")
    )
    .round(2)
    .reset_index()
)

monthly_trend_summary.to_csv(
    "../powerbi_data/monthly_trend_summary.csv",
    index=False
)

In [57]:
action_summary_table = (
    df["Allocation_Action"]
    .value_counts(normalize=True)
    .mul(100)
    .round(2)
    .reset_index()
)

action_summary_table.columns = ["Allocation_Action", "Percentage"]

action_summary_table.to_csv(
    "../powerbi_data/action_summary.csv",
    index=False
)